google_guest_agent/diagnostics.go

```diff
@@ -116,7 +120,11 @@ func (d *diagnosticsMgr) set() error {
 	if entry.Trace {
 		args = append(args, "-trace")
 	}
-
+	// If no existing running job, set it to 1 and block other requests
+	if !atomic.CompareAndSwapInt32(&isDiagnosticsRunningInt32, 0, 1) {
```

Comment:
maybe Andrew can give better suggestions(i am not an expert in go) but i think mutex is easy to use and read.

Reply:
A mutex would be easier to read, but you would lose the checking ability. You can also just use a channel which would be preferable:
https://play.golang.org/p/wiX0zeRSZnT

Reply:
I see. Found some discussions about Mutex vs Channel, [link1](https://github.com/golang/go/wiki/MutexOrChannel) & [link2](https://medium.com/mindorks/https-medium-com-yashishdua-synchronizing-states-using-mutex-vs-channel-in-go-25e646c83567). 
IIUC, mutex is preferable for this case, as we only need to change state without passing any data.
And I chose atomic is because it locks and update the gate variable in one function, it works like mutex but simpler.

Reply:
i still think mutex is easy to read but if you want to use CompareAndSwapInt32, you need to use it in the thread as well. i think CompareAndSwapInt32 include memory fence which make sure all processor gets the correct value.

Reply:
CompareAndSwapInt32 would be the simplest solution here (least code), and really I'm fine with it as this is a fairly simple lock, but as pointed out you need to do this atomicly at line 137 as well (use SwapInt32 to avoid a broken state).

Reply:
Got it. Done